### ISSUE TITLE
Settings registration of CLI commands

### ISSUE THREAD

**Issue by thoemmi** [spectreconsole/spectre.console#373]
Hi,

I've noticed that settings types for CLI commands are registered in the container:

https://github.com/spectreconsole/spectre.console/blob/3e2eea730bf0f6fe4a5e336faa312b526d351079/src/Spectre.Console/Cli/Internal/Extensions/TypeRegistrarExtensions.cs#L36

However, settings are instantiated using `Activator.CreateInstance`:

https://github.com/spectreconsole/spectre.console/blob/3e2eea730bf0f6fe4a5e336faa312b526d351079/src/Spectre.Console/Cli/Internal/Binding/CommandConstructorBinder.cs#L38

Therefore I don't see a reason why the settings types are registered at all.

In my case, I am using Microsoft's generic host, and have enabled [registration validation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.serviceprovideroptions.validateonbuild?view=dotnet-plat-ext-5.0). The validation fails because the container doesn't know how to resolve the registered `ExplainCommand.Settings` because of its c'tor parameters (`string[]? commands, bool? detailed, bool includeHidden`).

My suggestion is to remove the registration of commands' settings. If you agree, I'd be happy to provide a PR.

**Comment by phil-scott-78** [comment]
Good catch. I'm not seeing any reason for us to be registering those settings anymore. @patriksvensson might have some historical background on why it was being done though.

This would be a breaking behavior potentially, but I think anyone relying on this probably just got lucky because I'm not sure that behavior is documented. And moving forward probably good to rip it out now and break rather than build work arounds.

**Comment by thoemmi** [comment]
Sorry, I have to correct myself. `CommandBinder.Bind` decides based on some reflection logic, whether it uses `CommandConstructorBinder` (which in turn uses `Activator.CreateInstance`) or `CommandPropertyBinder.CreateSettings`, which uses the resolver.

I guess the same logic, that `CommandBinder.Bind` uses to decide about the instantiation, must also be used to decide whether the settings type must be registered in the first place.

**Comment by thoemmi** [comment]
It's even more complicated. `CommandBinder.Bind` first gets the parameters, and then checks if any constructor of the Settings type has parameters with the same name as the command line arguments. I am afraid that the same logic can't be used at the time of the registration :confused:

**Comment by phil-scott-78** [comment]
Dang. What if we approached it in a different direction?

Reading through `CommandBinder.Bind` it feels like the use case for the registrar being used for settings is if someone 1) isn't initializing any properties via the constructor AND 2) is injecting a service into the constructor. 

So, it seems like potentially a change would be making `CommandPropertyBinder.CreateSettings` 

1. Try the registrar. IF someone is doing the above use case then the onus would be on them to register their settings explicitly.
2. If that fails, then use `Activator.CreateInstance` against the default `ctor`. 

```csharp
private static CommandSettings CreateSettings(ITypeResolver resolver, Type settingsType)
{
    try
    {
        if (resolver.Resolve(settingsType) is CommandSettings settings)
        {
            return settings;
        }
    }
    catch
    {
        // ignored
    }


    if (Activator.CreateInstance(settingsType) is CommandSettings instance)
    {
        return instance;
    }

    throw CommandParseException.CouldNotCreateSettings(settingsType);
}
```

@patriksvensson, how does this feel?

**Comment by phil-scott-78** [comment]
Had some time tonight so I tried out my recommendation. With the exception of two unit tests named `Should_Register_Default_Command_Settings_When_Configuring_Application` and `Should_Register_Command_Settings_When_Configuring_Application` rightfully failing things seem to wire up nicely. I adjusted those two tests to explicitly inject the settings to make sure that was still possible and renamed them appropriately. 

https://github.com/spectreconsole/spectre.console/compare/main...phil-scott-78:bugfix/registration-of-settings

The big breaking change that I can think of is that someone out there might be injecting the settings into a service. I'm not sure that's a great design to begin with but with being able to explictly register they at least have a work around.

I'm not creating a PR yet because I'm sure I missed something, but hopefully this gets the conversation going.
